gadget/mountedfilesystem.go

```diff
@@ -186,3 +205,487 @@ func (m *MountedFilesystemWriter) writeVolumeContent(whereDir string, content *V
+	backupRoot := fsStructBackupPath(f.backupDir, f.ps)
+
+	for _, c := range f.ps.Content {
+		if err := f.updateVolumeContent(mount, &c, preserveInDst, backupRoot); err != nil {
+			return fmt.Errorf("cannot update content: %v", err)
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) updateDirectory(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName := f.entrySource(source)
+
+	if err := checkSrcIsDir(srcName); err != nil {
+		return fmt.Errorf("cannot update directory %q: %v", source, err)
+	}
+
+	fis, err := ioutil.ReadDir(srcName)
+	if err != nil {
+		return fmt.Errorf("cannot list directory entries: %v", err)
+	}
+
+	if !strings.HasSuffix(source, "/") {
+		// /foo -> /bar/ => /bar/foo
+		target = filepath.Join(target, filepath.Base(source))
+	}
+
+	for _, fi := range fis {
+		pSrc := filepath.Join(source, fi.Name())
+		pDst := filepath.Join(target, fi.Name())
+
+		update := f.updateOrSkipFile
+		if fi.IsDir() {
+			_, dstName, _ := f.entryNames(dstRoot, source, pDst, "")
+			if err := os.MkdirAll(dstName, 0755); err != nil {
+				return fmt.Errorf("cannot deploy directory: %v", err)
+			}
+
+			update = f.updateDirectory
+			pSrc += "/"
+		}
+		if err := update(dstRoot, pSrc, pDst, preserveInDst, backupDir); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (f *MountedFilesystemUpdater) updateOrSkipFile(dstRoot, source, target string, preserveInDst []string, backupDir string) error {
+	srcName, dstName, backupName := f.entryNames(dstRoot, source, target, backupDir)
+
+	if osutil.FileExists(dstName) {
+		if strutil.SortedListContains(preserveInDst, dstName) {
+			// file is to be preserved
+			return nil
+		}
+		if osutil.FileExists(backupName + ".same") {
+			// file is the same as current copy
+			return nil
+		}
+		if !osutil.FileExists(backupName + ".backup") {
+			// not preserver & different than the update, double
+			// check we have a backup
+			return fmt.Errorf("missing backup file for %v", dstName)
+		}
+	}
+
+	return writeFile(srcName, dstName, preserveInDst)
+}
+
+func (f *MountedFilesystemUpdater) updateVolumeContent(whereDir string, content *VolumeContent, preserveInDst []string, backupDir string) error {
+	if err := checkContent(content); err != nil {
+		return err
+	}
+
+	srcName, _, _ := f.entryNames(whereDir, content.Source, content.Target, backupDir)
+
+	if osutil.IsDirectory(srcName) || strings.HasSuffix(content.Source, "/") {
+		return f.updateDirectory(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	} else {
+		return f.updateOrSkipFile(whereDir, content.Source, content.Target, preserveInDst, backupDir)
+	}
+}
+
+// Backup analyzes a mounted filesystem and prepares a rollback state
+// information should the update be applied. The content of the filesystem is
+// processed, files that would be modified by the update are backed up, while
+// other files may be stamped to improve the later step of update process.
```

Comment:
this needs a much more detailed description of the algorithm, is not really that clear from code alone

Reply:
I'll extend the description of the whole thing.